src/cargo/ops/cargo_clean.rs

```diff
@@ -220,29 +222,134 @@ fn escape_glob_path(pattern: &Path) -> CargoResult<String> {
     Ok(glob::Pattern::escape(pattern))
 }
 
-fn rm_rf_glob(pattern: &Path, config: &Config) -> CargoResult<()> {
+fn rm_rf_glob(
+    pattern: &Path,
+    config: &Config,
+    progress: &mut impl CleaningProgressBar,
+) -> CargoResult<()> {
     // TODO: Display utf8 warning to user?  Or switch to globset?
     let pattern = pattern
         .to_str()
         .ok_or_else(|| anyhow::anyhow!("expected utf-8 path"))?;
     for path in glob::glob(pattern)? {
-        rm_rf(&path?, config)?;
+        rm_rf(&path?, config, progress)?;
     }
     Ok(())
 }
 
-fn rm_rf(path: &Path, config: &Config) -> CargoResult<()> {
-    let m = fs::symlink_metadata(path);
-    if m.as_ref().map(|s| s.is_dir()).unwrap_or(false) {
-        config
-            .shell()
-            .verbose(|shell| shell.status("Removing", path.display()))?;
-        paths::remove_dir_all(path).with_context(|| "could not remove build directory")?;
-    } else if m.is_ok() {
-        config
-            .shell()
-            .verbose(|shell| shell.status("Removing", path.display()))?;
-        paths::remove_file(path).with_context(|| "failed to remove build artifact")?;
+fn rm_rf(path: &Path, config: &Config, progress: &mut impl CleaningProgressBar) -> CargoResult<()> {
+    if fs::symlink_metadata(path).is_err() {
+        return Ok(());
     }
+
+    config
+        .shell()
+        .verbose(|shell| shell.status("Removing", path.display()))?;
+    progress.display_now()?;
+
+    for entry in walkdir::WalkDir::new(path).contents_first(true) {
+        let entry = entry?;
+        progress.on_clean()?;
+        if entry.file_type().is_dir() {
+            paths::remove_dir(entry.path()).with_context(|| "could not remove build directory")?;
+        } else {
+            paths::remove_file(entry.path()).with_context(|| "failed to remove build artifact")?;
+        }
+    }
+
     Ok(())
 }
+
+fn clean_entire_folder(path: &Path, config: &Config) -> CargoResult<()> {
+    let num_paths = walkdir::WalkDir::new(path).into_iter().count();
```

Comment:
This naively seems like it would have a large impact and slow things down because we have to walk everything before actually removing everything. I don't have a great sense for the practical impact of this though.

Reply:
An alternative to this I think would be to perform the walk once, collect a list of all paths to remove, then remove them all in a loop. That would mean we only have to walk once and should be similar in performance I think.